### PR TITLE
File PAL integration into mono

### PR DESCRIPTION
**Pull request by piconid**
integration of the the Unity PAL File api in to mono.

IL2CPP All tests katana with new source in il2cpp-dependencies
---------------------------------------------------------
https://katana.bf.unity3d.com/projects/IL2CPP/builders?il2cpp_branch=default&il2cpp-dependencies_branch=piconi-test-file-pal&il2cpp-userprojects_branch=default&il2cpp-xboxone_branch=default

Mono Katana
---------------
https://katana.bf.unity3d.com/projects/Mono%20Runtime%20and%20Classlibs/builders?boo_branch=unity-trunk&cecil_branch=unity-trunk&il2cpp_branch=default&krait-signal-handler_branch=master&mono_branch=unity-2017-02-file-pal&mono-build-deps_branch=default&unityscript_branch=unity-trunk

### CODE REVIEW

File: mono/metadata/w32file-unity.c [hunk 1]
```diff
@@ -1,282 +1,527 @@
...
+	int32_t resultAttributes = 0;
+	int32_t result;
+ 	const char* filename;
+
+	result = UnityPalDirectoryFindNextFile(handle, &filename, &resultAttributes);
+
+	find_data->dwFileAttributes = resultAttributes;
+	gunichar2 *utf16_basename;
+	glong bytes;
+	utf16_basename = g_utf8_to_utf16 (filename, -1, NULL, &bytes, NULL);
```
Comment (piconid) [inline mono/metadata/w32file-unity.c]: Memory leak, I need to fix

File: mono/metadata/w32file-unity.c [hunk 2]
```diff
@@ -1,282 +1,527 @@
...
+ 	{
+ 		mono_w32error_set_last(result);
+ 		return INVALID_HANDLE_VALUE;
+ 	}
+
+	find_data->dwFileAttributes = resultAttributes;
+
+	gunichar2 *utf16_basename;
+	glong bytes;
+	utf16_basename = g_utf8_to_utf16 (filename, -1, NULL, &bytes, NULL);
```
Comment (piconid) [inline mono/metadata/w32file-unity.c]: Memory leak I need to fix

File: mono/metadata/w32file-unity.c [hunk 3]
```diff
@@ -1,282 +1,529 @@
...
+	bytes *= 2;
+
+	memset (find_data->cFileName, '\0', (MAX_PATH * 2));
```
Comment (joncham) [inline mono/metadata/w32file-unity.c]: I am unclear who allocated this memory and how we know the length here?

File: mono/metadata/w32file-unity.c [hunk 4]
```diff
@@ -1,282 +1,529 @@
...
 {
-	g_assert(0 && "This function is not yet implemented for the Unity platform.");
-	return FALSE;
+	gboolean result;
+	*error = 0;
+	MONO_ENTER_GC_SAFE;
+	
+	gchar* palPath = mono_unicode_to_external(path);
+	gchar* palDest = mono_unicode_to_external(dest);
+    result =  UnityPalMoveFile(palPath, palDest, error);
```
Comment (joncham) [inline mono/metadata/w32file-unity.c]: Whitespace?

File: mono/metadata/w32file.c [hunk 5]
```diff
@@ -1052,13 +1057,23 @@ ves_icall_System_IO_MonoIO_SetFileTime (HANDLE handle, gint64 creation_time,
...
+
+#else
+
 	const FILETIME *creation_filetime;
 	const FILETIME *access_filetime;
 	const FILETIME *write_filetime;
-	MONO_ENTER_GC_SAFE;
 
-	*error=ERROR_SUCCESS;
-
```
Comment (joshpeterson) [inline mono/metadata/w32file.c]: Do these lines need to be moved above?

File: mono/metadata/w32file-unity.c [hunk 6]
```diff
@@ -1,282 +1,529 @@
...
+	return UnityPalGetFileType(handle);
 }
 
 gboolean
 mono_w32file_get_times (gpointer handle, FILETIME *create_time, FILETIME *access_time, FILETIME *write_time)
 {
-	g_assert(0 && "This function is not yet implemented for the Unity platform.");
-	return FALSE;
+	/* Not Supported in UnityPAL */
+	g_assert_not_reached();
```
Comment (joshpeterson) [inline mono/metadata/w32file-unity.c]: We should standardize a not support assert, I think. I was using this for the event code:

```
g_assertion_message("Named events are not supported by the Unity platform.");
```

I'm not sure if that is the best option though. Any other ideas?

File: mono/metadata/w32file-unity.c [hunk 7]
```diff
@@ -1,282 +1,529 @@
...
+	}
+
+	if (destinationBackupFileName != NULL)
+	{
+		destBackupPath = mono_unicode_to_external(destinationBackupFileName);
+	}
+
+	MONO_ENTER_GC_SAFE;
+
+	result =  UnityPalReplaceFile(sourcePath, destPath, destBackupPath, 0, error);
```
Comment (joshpeterson) [inline mono/metadata/w32file-unity.c]: Does `UnityPalReplaceFile` handle `NULL` values properly?

File: mono/metadata/w32file-unity.c [hunk 8]
```diff
@@ -1,282 +1,529 @@
...
-	g_assert(0 && "This function is not yet implemented for the Unity platform.");
-	return FALSE;
+	gboolean result;
+	*error = 0;
+
+	MONO_ENTER_GC_SAFE;
+	
+	gchar* palPath = mono_unicode_to_external(path);
+	gchar* palDest = mono_unicode_to_external(dest);
+	result = UnityPalCopyFile( palPath, palDest, overwrite, error);
```
Comment (joshpeterson) [inline mono/metadata/w32file-unity.c]: Nit: It looks like there is an extra space after the opening parenthesis here.

File: mono/metadata/w32file-unity.c [hunk 9]
```diff
@@ -1,282 +1,529 @@
...
 {
-	g_assert(0 && "This function is not yet implemented for the Unity platform.");
-	return 0;
+	gunichar2 *utf16_path;
+	glong count;
+	uintptr_t bytes;
+	int error = 0;
+
+	const char* palPath = UnityPalDirectoryGetCurrent(&error);
+    mono_w32error_set_last (error);
```
Comment (joshpeterson) [inline mono/metadata/w32file-unity.c]: Nit: More whitespace oddness.

File: mono/metadata/w32file-unity.c [hunk 10]
```diff
@@ -1,282 +1,529 @@
...
+	utf16_path = mono_unicode_from_external(palPath, &bytes);
+	count = (bytes / 2) + 1;
+	memcpy(buffer, utf16_path, length);
+	g_free(utf16_path);
+	g_free((void *)palPath);
```
Comment (joshpeterson) [inline mono/metadata/w32file-unity.c]: I don't think we need the cast to `void*` here.

File: mono/metadata/w32file.c (hunk 5 above)
Comment (piconid) [inline mono/metadata/w32file.c]: yep they are in correct place, the diff tool makes it look incorrect.

File: mono/metadata/w32file-unity.c (hunk 7 above)
Comment (piconid) [inline mono/metadata/w32file-unity.c]: yep, I doubled checked down in libil2cpp as well.

File: mono/metadata/w32file-unix.c [hunk 11]
```diff
@@ -3103,6 +3105,8 @@ mono_w32file_set_times(gpointer handle, const FILETIME *create_time, const FILET
 					 write_time));
 }
 
+#endif
```
Comment (joshpeterson) [inline mono/metadata/w32file-unix.c]: I've removed the code that tries to build the IL2CPP on Mono static libraries now, so we should be able to remove this #if wrapper.

**Pull request merged**